### PR TITLE
Support repeated values NAME=awsenv:/VALUE;NAME2=awsenv:/VALUE

### PR DESCRIPTION
**Pull request by derekbassett**
Allow support for multiple envvars pulling the same value in a configuration.

### CODE REVIEW

File: awsenv/helpers.go [hunk 1]
```diff
@@ -62,9 +51,9 @@ func chunk(size int, lst []string) [][]string {
 	return chunks
 }
 
-func pathmap(prefix string, env []string) map[string]string {
-	// param path -> env var name
-	m := make(map[string]string, len(env))
+// parseEnvironment takes the results of environ and converts it into a map of Environment Keys => Values
+func parseEnvironment(env []string) map[string]string {
```
Comment (kheppenstall) [inline awsenv/helpers.go:55]: Should we add unit tests for this given that `pathmap` was covered?

File: awsenv/replacer_test.go [hunk 2]
```diff
@@ -133,3 +184,87 @@ type mockParamsGetter func(context.Context, []string) (map[string]string, error)
...
+	for _, test := range tests {
+		r := Replacer{prefix: test.prefix}
+		got, want := r.filterPaths(test.input), test.want
+		if !reflect.DeepEqual(got, want) {
+			t.Errorf("filterPaths(%q) = %v, want %v", test.input, got, want)
+		}
+	}
+}
+
+func TestApplyPaths(t *testing.T) {
```
Comment (kheppenstall) [inline awsenv/replacer_test.go:220]: Should this be `TestReplacer_applyParamPathValues`?

File: awsenv/v1/v1.go [hunk 3]
```diff
@@ -4,17 +4,29 @@ package v1
...
+	"github.com/aws/aws-sdk-go/aws/request"
+	"github.com/aws/aws-sdk-go/aws/session"
 	"github.com/aws/aws-sdk-go/service/ssm"
+
 	"github.com/sendgrid/aws-env/awsenv"
 )
 
+// SSMGetParametersAPI defines the interface for the GetParameters function.
+// We use this interface to test the function using a mocked service.
+type SSMGetParametersAPI interface {
```
Comment (kheppenstall) [inline awsenv/v1/v1.go:17]: Why is tihs exported?

File: awsenv/v1/v1.go [hunk 4]
```diff
@@ -43,3 +55,15 @@ func (f *fetcher) GetParams(ctx context.Context, names []string) (map[string]str
 
 	return m, nil
 }
+
+func MustReplaceEnv() {
```
Comment (kheppenstall) [inline awsenv/v1/v1.go:59]: Missing a comment here on n exported func.

File: awsenv/v2/v2.go [hunk 5]
```diff
@@ -4,17 +4,27 @@ package v2
...
 
+	"github.com/aws/aws-sdk-go-v2/config"
 	"github.com/aws/aws-sdk-go-v2/service/ssm"
+
 	"github.com/sendgrid/aws-env/awsenv"
 )
 
+// SSMGetParametersAPI defines the interface for the GetParameters function.
+// We use this interface to test the function using a mocked service.
+type SSMGetParametersAPI interface {
```
Comment (kheppenstall) [inline awsenv/v2/v2.go:15]: Does tihs need to be exported?

File: awsenv/replacer_test.go [hunk 6]
```diff
@@ -133,3 +184,87 @@ type mockParamsGetter func(context.Context, []string) (map[string]string, error)
...
+			prefix:            "awsenv:",
+			src:               map[string]string{"x": "awsenv:/a", "y": "awsenv:/b"},
+			replaceWithValues: map[string]string{"/a": "A", "/b": "B"},
+			want:              map[string]string{"x": "A", "y": "B"},
+		},
+	}
+
+	for idx, test := range tests {
+		test := test
+		t.Run(fmt.Sprintf("test_%v", idx), func(t *testing.T) {
```
Comment (kheppenstall) [inline awsenv/replacer_test.go:261]: Should we add a `name` to the test struct so the failure message does not just include an index?

File: awsenv/replacer_test.go [hunk 7]
```diff
@@ -133,3 +184,87 @@ type mockParamsGetter func(context.Context, []string) (map[string]string, error)
...
+			r := &Replacer{prefix: test.prefix}
+			got, want := r.applyParamPathValues(test.src, test.replaceWithValues), test.want
+			if !reflect.DeepEqual(got, want) {
+				t.Errorf("applyPaths(%v, %v) -> %v, want %v",
```
Comment (kheppenstall) [inline awsenv/replacer_test.go:265]: `applyParamPathValues`

File: awsenv/replacer_test.go (hunk 2 above)
Comment (kheppenstall) [reply awsenv/replacer_test.go:220]: Missing `t.Parallel()` calls in here.

File: awsenv/replacer_test.go [hunk 8]
```diff
@@ -133,3 +184,87 @@ type mockParamsGetter func(context.Context, []string) (map[string]string, error)
 func (f mockParamsGetter) GetParams(ctx context.Context, paths []string) (map[string]string, error) {
 	return f(ctx, paths)
 }
+
+func TestFilterPaths(t *testing.T) {
```
Comment (kheppenstall) [inline awsenv/replacer_test.go:188]: `TestReplacer_filterPaths`?

Comment (kheppenstall) [reply awsenv/replacer_test.go:188]: Also missing `t.Parallel()`.

File: awsenv/replacer_test.go [hunk 9]
```diff
@@ -133,3 +184,87 @@ type mockParamsGetter func(context.Context, []string) (map[string]string, error)
...
+			prefix: "pre:",
+			input:  map[string]string{"X": "1", "Y": "pre:/y", "Z": "awsenv:/z"},
+			want:   []string{"/y"},
+		},
+	}
+
+	for _, test := range tests {
+		r := Replacer{prefix: test.prefix}
+		got, want := r.filterPaths(test.input), test.want
+		if !reflect.DeepEqual(got, want) {
```
Comment (kheppenstall) [inline awsenv/replacer_test.go:214]: Why are we using `reflect` for comparison here when we use packages like `/stretchr/testify` for assertions in other tests?

File: awsenv/replacer_test.go [hunk 10]
```diff
@@ -133,3 +184,87 @@ type mockParamsGetter func(context.Context, []string) (map[string]string, error)
...
+			want:   []string{"/z"},
+		},
+		{
+			prefix: "pre:",
+			input:  map[string]string{"X": "1", "Y": "pre:/y", "Z": "awsenv:/z"},
+			want:   []string{"/y"},
+		},
+	}
+
+	for _, test := range tests {
```
Comment (kheppenstall) [inline awsenv/replacer_test.go:211]: Is there a reason we are not  using our standard subtest pattern (i.e `t.Run(...` in this PR?

File: awsenv/replacer_test.go [hunk 11]
```diff
@@ -60,6 +78,39 @@ func TestReplacerMultiple(t *testing.T) {
 	require.Equal(t, want, env)
 }
 
+func TestReplacerMultipleSameValue(t *testing.T) {
```
Comment (kheppenstall) [inline awsenv/replacer_test.go:81]: `TestReplacer_ReplaceAll_MultipleSameValue`?

File: awsenv/replacer.go [hunk 12]
```diff
@@ -136,3 +160,20 @@ func fetch(ctx context.Context, ssm ParamsGetter, paths []string) (map[string]st
 
 	return dest, nil
 }
+
+// apply applies values from src keys translated through
```
Comment (kheppenstall) [inline awsenv/replacer.go:164]: `applyParamPathValues`

File: awsenv/replacer.go [hunk 13]
```diff
@@ -73,23 +73,47 @@ func (r *Replacer) ReplaceAll(ctx context.Context) error {
...
-	// env name -> env value
-	dest := make(map[string]string, len(pathvals))
-	translate(dest, pathvars, pathvals)
+	envvars = r.applyParamPathValues(envvars, pathvals)
+	return envvars, nil
+}
 
-	return dest, nil
+func (r *Replacer) filterPaths(envvars map[string]string) []string {
+	// param path
```
Comment (kheppenstall) [inline awsenv/replacer.go:104]: Not sure what this comment means.

File: awsenv/replacer_test.go (hunk 9 above)
Comment (derekbassett) [reply awsenv/replacer_test.go:214]: I didn't change that from the original code.

**Review by kheppenstall — APPROVED** [review]

**Pull request merged**